### ISSUE TITLE
Unexpected behavior on pruning when run optuna with my neural network

### ISSUE THREAD

**Issue by mikeaalv** [optuna/optuna#984]
Hey, Optuna seems to behave widely for me. The run code is like this :
```
study=optuna.create_study(direction="minimize",sampler=optuna.integration.CmaEsSampler(),pruner=optuna.pruners.SuccessiveHalvingPruner())
study.optimize(lambda trial: objective(trial,args,traindataloader,testdataloader),n_trials=100)
```

The objective function is like this: 

```
def objective(trial,args,traindataloader,testdataloader):
    n_layers=trial.suggest_int('n_layers',2,9)
    laryersize_ratio=trial.suggest_loguniform('layersize_ratio',0.5,8)
    ##create model
    model=models.__dict__['resnet2x_mlp'](ninput=args.ntheta,num_response=args.nspec,p=args.p,ncellscale=laryersize_ratio,x=n_layers)
    # model=torch.nn.DataParallel(model).cuda()
    model=torch.nn.DataParallel(model)
    if args.gpu_use==1:
        device=torch.device("cuda:0")#cpu
    else:
        device=torch.device("cpu")
    
    model.to(device)
    if args.optimizer=="sgd":
        optimizer=optim.SGD(model.parameters(),lr=args.learning_rate,momentum=args.momentum)
    elif args.optimizer=="adam":
        optimizer=optim.Adam(model.parameters(),lr=args.learning_rate)
    elif args.optimizer=="nesterov_momentum":
        optimizer=optim.SGD(model.parameters(),lr=args.learning_rate,momentum=args.momentum,nesterov=True)
    
    if args.scheduler=='step':
        scheduler=lr_scheduler.StepLR(optimizer,step_size=200,gamma=0.1)
    elif args.scheduler=='plateau':
        scheduler=lr_scheduler.ReduceLROnPlateau(optimizer,'min',factor=0.1)
    else:
        scheduler=None
    
    cudnn.benchmark=True
    ##model training
    for epoch in range(1,args.epochs+1):
        acctr=train(args,model,traindataloader,optimizer,epoch,device,args.ntime)
        acc1=test(args,model,testdataloader,device,args.ntime)
        # test(args,model,traindataloader,device,ntime) # to record the performance on training sample with model.eval()
        if scheduler is not None:
            if args.scheduler=='step':
                scheduler.step()
            elif args.scheduler=='plateau':
                scheduler.step(acc1)##not a good implementation, just try to do only train+test
        if epoch==1:
            best_acc1=acc1
            best_train_acc=acctr
        
        # is_best=acc1>best_acc1
        is_best=acc1<best_acc1
        is_best_train=acctr<best_train_acc
        # best_acc1=max(acc1,best_acc1)
        best_acc1=min(acc1,best_acc1)
        best_train_acc=min(acctr,best_train_acc)
        ## this set up cannot be fast and on time but there is no callback for real pytorch yet
        trial.report(acc1,epoch)

        # Handle pruning based on the intermediate value.
        if trial.should_prune():
            raise optuna.exceptions.TrialPruned()
    
    return acc1
```

The objective here is based on performance(MSE) on the 'test' set (I will have another set for final test). The MSE used for pruning (trial.report) is also the MSE output in the latter examples. The same MSE for the last epoch is returned by the objective function. I expect the best-selected model with minimal value on 'test' set. From the output I got, it doesn't behave as expected (the model with the smallest test set MSE wasn't selected as best model).

A few section of ouput file:
```
Train Epoch: 300 [4008690/4040000 (99%)]	Loss(per sample): 0.915393

Test set: Average loss (per sample): 0.9090

[I 2020-02-27 20:01:55,834] Finished trial#0 resulted in value: 0.9089573018888876. Current best value is 0.9089573018888876 with parameters: {'n_layers': 7, 'layersize_ratio': 1.5217470525102508}.
Train Epoch: 1 [0/4040000 (0%)]	Loss(per sample): 27054.690735
```

```
Train Epoch: 48 [4008690/4040000 (99%)]	Loss(per sample): 0.635707

Test set: Average loss (per sample): 0.6413

[I 2020-02-28 06:31:09,671] Setting status of trial#2 as TrialState.PRUNED. 
Train Epoch: 1 [0/4040000 (0%)]	Loss(per sample): 3671.041618
Train Epoch: 1 [49490/4040000 (1%)]	Loss(per sample): 466.115218
```

```
Test set: Average loss (per sample): 1.2290

[I 2020-02-28 23:05:26,226] Setting status of trial#99 as TrialState.PRUNED. 
Number of finished trials:  100
Best trial:
  Value:  0.9089573018888876
  Params: 
    n_layers: 7
    layersize_ratio: 1.5217470525102508
```

As you can see the best model wasn't recorded. Can you help see where I did wrong?

Thank you.

Best.
Yue

Regarding the questions: 
1. I haven't try to turn off pruning. I can do it but it will take a few days as each full run takes a few hours. 
2. I'm dealing with a regression problem and there is MSE to be minimized for both neural network and hyperparameter tuning. The acctr or acc is generic word for returned value from train or test.(In our case, it is MSE not accuracy).

**Comment by toshihikoyanase** [comment]
Thank you for your detailed report.

I guess the overfitting occurs during the training. The score of Trial#2 at 48th epoch is lower than that of Trial#0 at 300th epoch, but Trial#0 might mark better score at 48th epoch. If so, it is expected that `SuccessiveHalvingPruner` prunes Trial#2.

I think you can check overfitting with the intermediate result plot. You can easily plot the result using [optuna.visualization.plot_intermediate_values](https://optuna.readthedocs.io/en/latest/reference/visualization.html#optuna.visualization.plot_intermediate_values).

The following is an example of the intermediate result plot. 
![image](https://user-images.githubusercontent.com/3255979/76080783-0ab7ea80-5feb-11ea-8aa0-47506dcdafa4.png)

If the `SuccessiveHalvingPruner` pruned the better results at a certain epoch (e.g., 48th epoch), it may have a bug.

**Comment by mikeaalv** [comment]
I checked the case you mentioned. I turns out the minimum MSE at trial 0 is indeed slightly lower (~0.907) but not as low as the end test MSE at trial 1 (~0.64). The MSE at trial 0 at epoch 48 is higher (0.9475). You can check it from this file https://www.dropbox.com/s/0vbprux0veenesn/testmodel.sapelo2.out?dl=0

Also as you mentioned the case of overfitting, that definitely can happen at any train (test MSE get bigger). Can I "automatically" deal with reasonably? Do you think trial.report best test MSE is a better choice than trial.report every epoch end MSE.

**Comment by toshihikoyanase** [comment]
I'm sorry for the delayed response. I plotted the intermediate values of each trial and found out that pruned trials have higher intermediate values than trial#0 or trial#1. So, I think that this is the expected behavior. Please take a look at the following figure:
![plot_intermediate](https://user-images.githubusercontent.com/3255979/77970542-698d2d00-7327-11ea-8827-1064de3f2aaf.png)

If you feel 'SuccessiveHalvingPruner' prunes too many trials, please think of [`HyperbandPruner`](https://optuna.readthedocs.io/en/latest/reference/pruners.html#optuna.pruners.HyperbandPruner) which can tune pruning behavior automatically.

**Comment by mikeaalv** [comment]
Thank you very much. Sorry forgot to check the minimum of trial #1, which turns out really small (0.492). Trial #2 did turn back overfitting and the final result is worse than 1.

**Comment by toshihikoyanase** [comment]
@mikeaalv Thank you for your reply. 
I think this issue was resolved, so let me close it. Please feel free to re-open it or create a new PR if you have further questions or comments.
